examples/src/main/scala/shapeless/examples/deepGeneric.scala

```diff
@@ -0,0 +1,98 @@
+package shapeless.examples
+
+import shapeless._, test._
+
+trait DeepGeneric[T] {
+  type Repr
+
+  def to(t: T): Repr
+  def from(t: Repr): T
+}
+
+trait LowPriorityDeepGeneric {
+  implicit def hlistHeadNotGeneric[H, T <: HList, TR <: HList](implicit dgt: Lazy[DeepGeneric.Aux[T, TR]]): DeepGeneric.Aux[H :: T, H :: TR] =
+    new DeepGeneric[H :: T] {
+      type Repr = H :: TR
+      def to(hlist: H :: T): Repr = hlist match {
+        case h :: t => h :: dgt.value.to(t)
+      }
+      def from(r: Repr): H :: T = r match {
+        case h :: t => h :: dgt.value.from(t)
+      }
+    }
+}
+
+object DeepGeneric extends LowPriorityDeepGeneric {
+  type Aux[T, Repr0] = DeepGeneric[T] { type Repr = Repr0 }
+
+  implicit val hnilDeepGeneric: DeepGeneric.Aux[HNil, HNil] = new DeepGeneric[HNil] {
+    type Repr = HNil
+    def to(h: HNil): HNil = h
+    def from(h: HNil): HNil = h
+  }
+
+  implicit def hlistDeepGeneric[H, HG, T <: HList, TR <: HList, HR](
+    implicit hg: Generic.Aux[H, HG],
+    hgen: Lazy[DeepGeneric.Aux[HG, HR]],
+    tgen: Lazy[DeepGeneric.Aux[T, TR]]
+  ): DeepGeneric.Aux[H :: T, HR :: TR] = new DeepGeneric[H :: T] {
```

Comment:
This fails to compile with `ambiguous implicit values.  both method hlistDeepGeneric and method hlistHeadNotGeneric match expected type`.  I would have thought that `hlistDeepGeneric` would get prioritized.

Perplexingly, if I add the constraint `HR <: HList` then the code compiles.  Since my goal is to include coproducts, I don't want to do this.  

Any ideas on what I've done wrong?